src/mumble/GlobalShortcut_win.cpp

```diff
@@ -130,576 +52,356 @@ GlobalShortcutWin::~GlobalShortcutWin() {
+			// Right button pressed
+			button = 2;
+			down   = true;
+			break;
+		case RI_MOUSE_BUTTON_2_UP:
+			// Right button released
+			button = 2;
+			down   = false;
+			break;
+		case RI_MOUSE_BUTTON_3_DOWN:
+			// Middle button pressed
+			button = 3;
+			down   = true;
+			break;
+		case RI_MOUSE_BUTTON_3_UP:
+			// Middle button released
+			button = 3;
+			down   = false;
+			break;
+		case RI_MOUSE_BUTTON_4_DOWN:
+			// Side button 1 pressed
+			button = 4;
+			down   = true;
+			break;
+		case RI_MOUSE_BUTTON_4_UP:
+			// Side button 1 released
+			button = 4;
+			down   = false;
+			break;
+		case RI_MOUSE_BUTTON_5_DOWN:
+			// Side button 2 pressed
+			button = 5;
+			down   = true;
+			break;
+		case RI_MOUSE_BUTTON_5_UP:
+			// Side button 2 released
+			button = 5;
+			down   = false;
+			break;
+		default:
+			return;
 	}
-	return CallNextHookEx(gsw->hhKeyboard, nCode, wParam, lParam);
-}
 
-LRESULT CALLBACK GlobalShortcutWin::HookMouse(int nCode, WPARAM wParam, LPARAM lParam) {
-	GlobalShortcutWin *gsw = static_cast< GlobalShortcutWin * >(engine);
-	MSLLHOOKSTRUCT *mouse  = reinterpret_cast< MSLLHOOKSTRUCT * >(lParam);
-
-	if (nCode >= 0) {
-		bool suppress = false;
-		UINT msg      = wParam;
-		// Convert the hooked Windows mouse message into a DirectInput
-		// button index, and store the pressed state of the button.
-		bool down        = false;
-		unsigned int btn = 0;
-		switch (msg) {
-			case WM_LBUTTONDOWN:
-				down = true;
-			case WM_LBUTTONUP:
-				btn = 3;
-				break;
-			case WM_RBUTTONDOWN:
-				down = true;
-			case WM_RBUTTONUP:
-				btn = 4;
-				break;
-			case WM_MBUTTONDOWN:
-				down = true;
-			case WM_MBUTTONUP:
-				btn = 5;
-				break;
-			case WM_XBUTTONDOWN:
-				down = true;
-			case WM_XBUTTONUP:
-				btn = 5 + (mouse->mouseData >> 16);
-			default:
-				break;
-		}
-		if (btn > 0) {
-			suppress = handleMouseMessage(btn, down);
-			if (suppress) {
-				return 1;
-			}
-		}
-	}
-	return CallNextHookEx(gsw->hhMouse, nCode, wParam, lParam);
+	const QList< QVariant > ql{ QVariant::fromValue(DeviceTypes::MOUSE), button };
+	handleButton(ql, down) && button != 1;
 }
 
-BOOL CALLBACK GlobalShortcutWin::EnumDeviceObjectsCallback(LPCDIDEVICEOBJECTINSTANCE lpddoi, LPVOID pvRef) {
-	InputDevice *id             = static_cast< InputDevice * >(pvRef);
-	QString name                = QString::fromUtf16(reinterpret_cast< const ushort * >(lpddoi->tszName));
-	id->qhNames[lpddoi->dwType] = name;
-
-	if (Global::get().s.bDirectInputVerboseLogging) {
-		qWarning("GlobalShortcutWin: EnumObjects: device %s %s object 0x%.8lx %s",
-				 qPrintable(QUuid(id->guid).toString()), qPrintable(id->name),
-				 static_cast< unsigned long >(lpddoi->dwType), qPrintable(name));
+GlobalShortcutWin::Devices::iterator GlobalShortcutWin::addDevice(const HANDLE deviceHandle) {
+	RID_DEVICE_INFO deviceInfo;
+	UINT size = sizeof(deviceInfo);
+	if (GetRawInputDeviceInfo(deviceHandle, RIDI_DEVICEINFO, &deviceInfo, &size) <= 0) {
```

Comment:
Why do we init `size` with a value and then pass it as a pointer? Is this an optional parameter in the API?

Reply:
It's a required parameter, the reason why it's received as a pointer is because the function can write the required buffer size to it.

From https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-getrawinputdeviceinfoa:
> If successful, this function returns a non-negative number indicating the number of bytes copied to _pData_.
>
> If _pData_ is not large enough for the data, the function returns -1. If _pData_ is __NULL__, the function returns a value of zero. In both of these cases, _pcbSize_ is set to the minimum size required for the _pData_ buffer.

Reply:
That doesn't explain why we initialize the variable to the given value though, does it?

Reply:
We're telling the function how big our buffer is:

![](https://user-images.githubusercontent.com/5897523/116124628-d149af00-a6c4-11eb-8bce-32b8f96881db.png)

Reply:
So this is an input _and_ an output variable? Man that is some ugly API :eyes:

Reply:
Two separate functions would've been better, because with `RID_DEVICE_INFO` the variable is just used an input.